### DISCUSSION TITLE
Webserver and OTA Update  - only one is working... why?

### DISCUSSION THREAD

**Discussion by Bordman-ger** [JAndrassy/WiFiEspAT#97] [Q&A] [answered]
Hi, finaly I got the board Wemos Mega +WiFi R3 Module ATmega2560+ESP8266 working , was able to activate AT Firmware 1.7.5. and got the example sketches WIFIServer and the OTA WIFI  working seperatly, now I am trying to combine these two, but whatever is started first is blocking the other...
why? How to get both running in parallel? 

```  
// waiting for connection to Wifi network set with the SetupWiFiConnection sketch
  Serial.println("Waiting for connection to WiFi");
  while (WiFi.status() != WL_CONNECTED) {
    delay(1000);
    Serial.print('.');
  }
  Serial.println();

      // start the WiFi OTA library with internal (flash) based storage
  ArduinoOTA.begin(WiFi.localIP(), "Arduino", "password", InternalStorage);


  server.begin();

  IPAddress ip = WiFi.localIP();
  printWifiStatus();
  Serial.println();
  Serial.println("Connected to WiFi network.6");
  Serial.print("To access the server, enter \"http://");
  Serial.print(ip);
  Serial.println("/\" in web browser.");

    // start the WiFi OTA library with internal (flash) based storage
 //ArduinoOTA.begin(WiFi.localIP(), "Arduino", "password", InternalStorage);
```

IF OTA is started first, Webserver fails...  and vice versa...
Full sketch: 

```
/*
  WiFi Web Server
 A simple web server that shows the value of the analog input pins.
 created 13 July 2010
 by dlf (Metodo2 srl)
 modified 31 May 2012
 by Tom Igoe
 modified in Jul 2019 for WiFiEspAT library
 by Juraj Andrassy https://github.com/jandrassy
 */

#include <WiFiEspAT.h>
#include <ArduinoOTA.h>

#define AT_BAUD_RATE 115200

WiFiServer server(80);

void setup() {

  Serial.begin(115200);
//  while (!Serial);

  Serial3.begin(AT_BAUD_RATE);
  WiFi.init(Serial3);

  if (WiFi.status() == WL_NO_SHIELD) {
    Serial.println("Communication with WiFi module failed!");
    // don't continue
    while (true);
  }

  // waiting for connection to Wifi network set with the SetupWiFiConnection sketch
  Serial.println("Waiting for connection to WiFi");
  while (WiFi.status() != WL_CONNECTED) {
    delay(1000);
    Serial.print('.');
  }
  Serial.println();

      // start the WiFi OTA library with internal (flash) based storage
  ArduinoOTA.begin(WiFi.localIP(), "Arduino", "password", InternalStorage);


  server.begin();

  IPAddress ip = WiFi.localIP();
  printWifiStatus();
  Serial.println();
  Serial.println("Connected to WiFi network.6");
  Serial.print("To access the server, enter \"http://");
  Serial.print(ip);
  Serial.println("/\" in web browser.");

    // start the WiFi OTA library with internal (flash) based storage
 //ArduinoOTA.begin(WiFi.localIP(), "Arduino", "password", InternalStorage);

 
}

void loop() {
  // check for WiFi OTA updates
  ArduinoOTA.poll();

  WiFiClient client = server.available();
  if (client) {
    IPAddress ip = client.remoteIP();
    Serial.print("new client ");
    Serial.println(ip);

    while (client.connected()) {
      if (client.available()) {
        String line = client.readStringUntil('\n');
        line.trim();
        Serial.println(line);

        // if you've gotten to the end of the HTTP header (the line is blank),
        // the http request has ended, so you can send a reply
        if (line.length() == 0) {
          // send a standard http response header
          client.println("HTTP/1.1 200 OK");
          client.println("Content-Type: text/html");
          client.println("Connection: close");  // the connection will be closed after completion of the response
          client.println("Refresh: 5");  // refresh the page automatically every 5 sec
          client.println();
          client.println("<!DOCTYPE HTML>");
          client.println("<html>");
       //   client.println(__FILE__ " " __DATE__ " " __TIME__);
          // output the value of analog input pins
          for (int analogChannel = 0; analogChannel < 4; analogChannel++) {
            int sensorReading = analogRead(analogChannel);
            client.print("analog input ");
            client.print(analogChannel);
            client.print(" is ");
            client.print(sensorReading);
            client.println("<br />");
          }
          client.println("</html>");
          client.flush();
          break;
        }
      }
    }

    // close the connection:
    client.stop();
    Serial.println("client disconnected");
  }
}
void printWifiStatus() {

  // print the SSID of the network you're attached to:
  char ssid[33];
  WiFi.SSID(ssid);
  Serial.print("SSID: ");
  Serial.println(ssid);

  // print the BSSID of the network you're attached to:
  uint8_t bssid[6];
  WiFi.BSSID(bssid);
  Serial.print("BSSID: ");
  printMacAddress(bssid);

  uint8_t mac[6];
  WiFi.macAddress(mac);
  Serial.print("MAC: ");
  printMacAddress(mac);

  // print your board's IP address:
  IPAddress ip = WiFi.localIP();
  Serial.print("IP Address: ");
  Serial.println(ip);

  // print the received signal strength:
  long rssi = WiFi.RSSI();
  Serial.print("signal strength (RSSI):");
  Serial.print(rssi);
  Serial.println(" dBm");
}

void printMacAddress(byte mac[]) {
  for (int i = 5; i >= 0; i--) {
    if (mac[i] < 16) {
      Serial.print("0");
    }
    Serial.print(mac[i], HEX);
    if (i > 0) {
      Serial.print(":");
    }
  }
  Serial.println();
}
```

**Comment by JAndrassy** [comment]
because standard AT firmware supports only one server
https://github.com/JAndrassy/WiFiEspAT#limitations

**Reply by Bordman-ger** [reply]
got it - well obviously I did not read that part - or dif not understand that OTA will be a second listening server...  )-: 

OK what are the easiest options for me? OTA Pull from a server? modified AT firmware? other? 
Obviousli it does not make sense to have OTA updates without any feedback from the device - divice will be lokated in the attic...

Thanks for suggestions...

Uwe

**Comment by JAndrassy** [comment]
If you don't need UDP then Jiri Bilek's firmware is the best option. I use it

**Comment by Bordman-ger** [comment]
Thank You... just short question - not sure if something went wrong or if it works as designed...
Wifi Connect worked - runing your Sketch for persistand Wifi worked as well but seems not to store the data? 
Trying to get the Web server running --- remains in Weiting to connect to WiFi...
Is Persisten setting still working or not supported with the modified AT firmware?

**Comment by JAndrassy** [comment]
Jiri Bilek's firmware? use esp8266 Arduino core 2.7.4 or for 3+ patch the fw source with https://github.com/JiriBilek/ESP_ATMod/commit/5cfdcb53e7675eaff849f1f54e52fb01c7537894

**Comment by Bordman-ger** [comment]
Perfect... sorry I did not find these comments, was looking for Discussions... but not the bug reports...
so afirst I did the patch and now it looks better, OTA is working if Ssrial is started and closed before OTA... okay... 
Web server is also startet but after 3 refrshs the connection fails - so it looks like all 5 possible connections will be used instead reusing the second one... would it be better to stay with 2.7.4? I made the switch to 2.7.4.  same behaviour...
Another thing I want to get confirmed... writing the sketch I used the  board manager with generic ESP8266 mdule - do I need to adopt settings? do I need to change the flash size? I assume the Board has 32 Mbits, your script is using 2 Mbits, but there are different OTA settings available... what is best to use? First I tried with default, no I changeed to 2MB FS 1MB OTA 512 KB, but looks pretty the  same
and the log is showing outo detect..
`esptool.py v2.8
Serial port COM7
Connecting....
Chip is ESP8266EX
Features: WiFi
Crystal is 26MHz
MAC: 98:f4:ab:e6:10:a2
Uploading stub...
Running stub...
Stub running...
Configuring flash size...
Auto-detected Flash size: 4MB
Flash params set to 0x0340
Compressed 457968 bytes to 334626...
Wrote 457968 bytes (334626 compressed) at 0x00000000 in 29.6 seconds (effective 123.6 kbit/s)...
Hash of data verified.

Leaving...
Hard resetting via RTS pin...
`

@JAndrassy one last thing for now... is there a way to pay your coffee with paypal?

**Comment by JAndrassy** [comment]
> Web server is also startet but after 3 refrshs the connection fails - so it looks like all 5 possible connections will be used instead reusing the second one... would it be better to stay with 2.7.4? I made the switch to 2.7.4. same behaviour...

with the WebServer example?

> different OTA settings available... what is best to use?

the fw requires FS, but doesn't use OTA (for fw). for a 4 MB flash any settings with some FS is OK. Just use always the same setting to preserve saved settings of the fw. 

>is there a way to pay your coffee with paypal?

I have PayPal for my gmail address juraj.andrassy@gmail.com

**Comment by Bordman-ger** [comment]
Hi, with  the modified Webserver from above, but I can try with your example again...
I see: my use of OTA is related to the Arduino, not to the ESP.. so not relevant

I just tried to start the debug mode... first block is with working refresh of browser, second is failing...
`

esp> 0,CONNECT ...processed
esp> +IPD,0,394 ...processed
esp INFO: incoming linkId 0
esp INFO: BuffManager returned buff.stream at 0 for server at port 80 for linkId 0
esp INFO: status of link 0
esp> AT+CIPSTATUS ...sent
esp> STATUS:3 ...matched
esp> +CIPSTATUS:0,"TCP","192.168.13.50",21357,80,0 ...matched
esp> OK ...matched
new client 192.168.13.50
esp INFO: get data on link 0
esp> AT+CIPRECVDATA=0,64 ...sent
esp> +CIPRECVDATA,64 ...matched
esp> OK ...matched
esp INFO: 	got 64 bytes on link 0
GET / HTTP/1.1
Host: 192.168.17.100
esp INFO: get data on link 0
esp> AT+CIPRECVDATA=0,64 ...sent
esp> +CIPRECVDATA,64 ...matched
esp> OK ...matched
esp INFO: 	got 64 bytes on link 0
esp INFO: get data on link 0
esp> AT+CIPRECVDATA=0,64 ...sent
esp> +CIPRECVDATA,64 ...matched
esp> OK ...matched
esp INFO: 	got 64 bytes on link 0
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:107.0) Gecko/20100101 Firefox/107.0
esp INFO: get data on link 0
esp> AT+CIPRECVDATA=0,64 ...sent
esp> +CIPRECVDATA,64 ...matched
esp> OK ...matched
esp INFO: 	got 64 bytes on link 0
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,*/*;q=0.8
esp INFO: get data on link 0
esp> AT+CIPRECVDATA=0,64 ...sent
esp> +CIPRECVDATA,64 ...matched
esp> OK ...matched
esp INFO: 	got 64 bytes on link 0
Accept-Language: de-DE,en-US;q=0.7,en;q=0.3
Accept-Encoding: gzip, deflate
DNT: 1
esp INFO: get data on link 0
esp> AT+CIPRECVDATA=0,64 ...sent
esp> +CIPRECVDATA,64 ...matched
esp> OK ...matched
esp INFO: 	got 64 bytes on link 0
Connection: keep-alive
Upgrade-Insecure-Requests: 1
esp INFO: get data on link 0
esp> AT+CIPRECVDATA=0,64 ...sent
esp> +CIPRECVDATA,10 ...matched
esp> OK ...matched
esp INFO: 	got 10 bytes on link 0
Cache-Control: max-age=0

esp INFO: send data on link 0
esp> AT+CIPSEND=0,64 ...sent
esp> OK ...ignored
esp> > ...matched
esp> Recv 64 bytes ...matched
esp> SEND OK ...matched
esp INFO: 	sent 64 bytes on link 0
esp INFO: send data on link 0
esp> AT+CIPSEND=0,64 ...sent
esp> OK ...ignored
esp> > ...matched
esp> Recv 64 bytes ...matched
esp> SEND OK ...matched
esp INFO: 	sent 64 bytes on link 0
esp INFO: send data on link 0
esp> AT+CIPSEND=0,64 ...sent
esp> OK ...ignored
esp> > ...matched
esp> Recv 64 bytes ...matched
esp> SEND OK ...matched
esp INFO: 	sent 64 bytes on link 0
esp INFO: send data on link 0
esp> AT+CIPSEND=0,64 ...sent
esp> OK ...ignored
esp> > ...matched
esp> Recv 64 bytes ...matched
esp> SEND OK ...matched
esp INFO: 	sent 64 bytes on link 0
esp INFO: send data on link 0
esp> AT+CIPSEND=0,64 ...sent
esp> OK ...ignored
esp> > ...matched
esp> Recv 64 bytes ...matched
esp> SEND OK ...matched
esp INFO: 	sent 64 bytes on link 0
esp INFO: AP query
esp> AT+CWJAP? ...sent
esp> +CWJAP:"PinheadWorld-IOT","7e:8a:20:4c:03:47",11,-43 ...matched
esp> OK ...matched
esp INFO: send data on link 0
esp> AT+CIPSEND=0,64 ...sent
esp> OK ...ignored
esp> > ...matched
esp> Recv 64 bytes ...matched
esp> SEND OK ...matched
esp INFO: 	sent 64 bytes on link 0
esp INFO: send data on link 0
esp> AT+CIPSEND=0,12 ...sent
esp> OK ...ignored
esp> > ...matched
esp> Recv 12 bytes ...matched
esp> SEND OK ...matched
esp INFO: 	sent 12 bytes on link 0
esp INFO: close link 0
esp> AT+CIPCLOSE=0 ...sent
esp> 0,CLOSED ...processed
esp INFO: closed linkId 0
esp> OK ...matched
client disconnected



esp> 0,CONNECT ...processed
esp> +IPD,0,394 ...processed
esp INFO: incoming linkId 0
esp INFO: BuffManager returned buff.stream at 0 for server at port 65280 for linkId 0
esp INFO: get data on link 0
esp> AT+CIPRECVDATA=0,64 ...sent
esp> +CIPRECVDATA,64 ...matched
esp> OK ...matched
esp INFO: 	got 64 bytes on link 0
esp INFO: get data on link 0
esp> AT+CIPRECVDATA=0,64 ...sent
esp> +CIPRECVDATA,64 ...matched
esp> OK ...matched
esp INFO: 	got 64 bytes on link 0
esp INFO: get data on link 0
esp> AT+CIPRECVDATA=0,64 ...sent
esp> +CIPRECVDATA,64 ...matched
esp> OK ...matched
esp INFO: 	got 64 bytes on link 0
esp INFO: get data on link 0
esp> AT+CIPRECVDATA=0,64 ...sent
esp> +CIPRECVDATA,64 ...matched
esp> OK ...matched
esp INFO: 	got 64 bytes on link 0
esp INFO: get data on link 0
esp> AT+CIPRECVDATA=0,64 ...sent
esp> +CIPRECVDATA,64 ...matched
esp> OK ...matched
esp INFO: 	got 64 bytes on link 0
esp INFO: get data on link 0
esp> AT+CIPRECVDATA=0,64 ...sent
esp> +CIPRECVDATA,64 ...matched
esp> OK ...matched
esp INFO: 	got 64 bytes on link 0
esp INFO: get data on link 0
esp> AT+CIPRECVDATA=0,64 ...sent
esp> +CIPRECVDATA,10 ...matched
esp> OK ...matched
esp INFO: 	got 10 bytes on link 0
esp INFO: sync
esp> AT+CIPSTATUS ...sent
esp> STATUS:3 ...matched
esp> +CIPSTATUS:0,"TCP","192.168.13.50",21359,80,0 ...matched
esp> OK ...end of list
esp> AT+CIPRECVLEN? ...sent
esp> +CIPRECVLEN:0,0,0,0,0 ...matched
esp> OK ...matched
esp INFO: send data on link 0
esp> AT+CIPSEND=0,64 ...sent
esp> OK ...ignored
esp> > ...matched
esp> Recv 64 bytes ...matched
esp> SEND OK ...matched
esp INFO: 	sent 64 bytes on link 0
esp INFO: send data on link 0
esp> AT+CIPSEND=0,35 ...sent
esp> OK ...ignored
esp> > ...matched
esp> Recv 35 bytes ...matched
esp> SEND OK ...matched
esp INFO: 	sent 35 bytes on link 0
esp INFO: close link 0
esp> AT+CIPCLOSE=0 ...sent
esp> 0,CLOSED ...processed
esp INFO: closed linkId 0
esp> UNLINK ...processed
esp> ERROR ...UNLINK is OK
`

**Comment by Bordman-ger** [comment]
Switched back to your Demo - is working fine!!!
so something is wrong in my sketch...??? in combination wit OTA? 

here is my sketch again... with some improvements for debugging...
```
/*
  WiFi Web Server
  A simple web server that shows the value of the analog input pins.
  created 13 July 2010
  by dlf (Metodo2 srl)
  modified 31 May 2012
  by Tom Igoe
  modified in Jul 2019 for WiFiEspAT library
  by Juraj Andrassy https://github.com/jandrassy
*/

#include <WiFiEspAT.h>
#include <ArduinoOTA.h>

#define AT_BAUD_RATE 115200

WiFiServer server(80);
int clientnum = 0;

void setup() {

  Serial.begin(115200);
  //  while (!Serial);

  Serial3.begin(AT_BAUD_RATE);
  WiFi.init(Serial3);

  if (WiFi.status() == WL_NO_SHIELD) {
    Serial.println("Communication with WiFi module failed!");
    // don't continue
    while (true);
  }

  // waiting for connection to Wifi network set with the SetupWiFiConnection sketch
  Serial.println("Waiting for connection to WiFi");
  while (WiFi.status() != WL_CONNECTED) {
    delay(1000);
    Serial.print('.');
  }
  Serial.println();

  // start the WiFi OTA library with internal (flash) based storage
  ArduinoOTA.begin(WiFi.localIP(), "Arduino", "password", InternalStorage);


  server.begin();

  IPAddress ip = WiFi.localIP();
  printWifiStatus();
  Serial.println();
  Serial.println("Connected to WiFi network.");
  Serial.print("To access the server, enter \"http://");
  Serial.print(ip);
  Serial.println("/\" in web browser.");

  // start the WiFi OTA library with internal (flash) based storage
  //   ArduinoOTA.begin(WiFi.localIP(), "Arduino", "password", InternalStorage);


}

void loop() {
  // check for WiFi OTA updates
  ArduinoOTA.poll();

  WiFiClient client = server.available();
  if (client) {
    clientnum++;
    IPAddress ip = client.remoteIP();
    Serial.print("new client ");
    Serial.println(ip);

    while (client.connected()) {
      if (client.available()) {
        String line = client.readStringUntil('\n');
        line.trim();
        Serial.println(line);

        // if you've gotten to the end of the HTTP header (the line is blank),
        // the http request has ended, so you can send a reply
        if (line.length() == 0) {
          // send a standard http response header
          client.println("HTTP/1.1 200 OK");
          client.println("Content-Type: text/html");
          client.println("Connection: close");  // the connection will be closed after completion of the response
          client.println("Refresh: 10");  // refresh the page automatically every 5 sec
          client.println();
          client.println("<!DOCTYPE HTML>");
          client.println("<html>");
          client.println(__FILE__ " " __DATE__ " " __TIME__);
          client.println("<br />");
          // output the value of analog input pins
          for (int analogChannel = 0; analogChannel < 4; analogChannel++) {
            int sensorReading = analogRead(analogChannel);
            client.print("analog input ");
            client.print(analogChannel);
            client.print(" is ");
            client.print(sensorReading);
            client.println("<br />");
          }
          client.print(" Client Number ");
          client.print(clientnum);
          client.println("<br />");
          client.print(" Version 2 ");
          client.println("<br />");
          long rssi = WiFi.RSSI();
          client.print("signal strength (RSSI):");
          client.print(rssi);
          client.println(" dBm");
          client.println("<br />"); 



        client.println("</html>");
        client.flush();
        break;
      }
    }
  }

  // close the connection:
  client.stop();
  Serial.println("client disconnected");
}
}
void printWifiStatus() {

  // print the SSID of the network you're attached to:
  char ssid[33];
  WiFi.SSID(ssid);
  Serial.print("SSID: ");
  Serial.println(ssid);

  // print the BSSID of the network you're attached to:
  uint8_t bssid[6];
  WiFi.BSSID(bssid);
  Serial.print("BSSID: ");
  printMacAddress(bssid);

  uint8_t mac[6];
  WiFi.macAddress(mac);
  Serial.print("MAC: ");
  printMacAddress(mac);

  // print your board's IP address:
  IPAddress ip = WiFi.localIP();
  Serial.print("IP Address: ");
  Serial.println(ip);

  // print the received signal strength:
  long rssi = WiFi.RSSI();
  Serial.print("signal strength (RSSI):");
  Serial.print(rssi);
  Serial.println(" dBm");
}

void printMacAddress(byte mac[]) {
  for (int i = 5; i >= 0; i--) {
    if (mac[i] < 16) {
      Serial.print("0");
    }
    Serial.print(mac[i], HEX);
    if (i > 0) {
      Serial.print(":");
    }
  }
  Serial.println();
}
```

**Comment by Bordman-ger** [comment]
have startet with you example and juste added OTA...

As soon as I  add  ArduinoOTA.poll(); in the loop, the issue is starting...

**Comment by JAndrassy** [comment]
but the second block is OTA. and the output doesn't show a fail. `UNLINK` is OK.
debug slows the library down . OTA will not work reliably with debug on.
for OTA I recommend to change the baud rate for AT to 250000 baud.

**Comment by Bordman-ger** [comment]
Hi, I tried with debug off and 250000, and now also with 500000, no change... 

MAybe I should cler ESP and write the Modiefied AT Sketch again? what is the best procedure to erease all? Or is this noit relevant when I am updating wito IDE and sketch? I am now on ESP8266 Version 2.6.4, modified Jiri Bilek's firmware , Arduino Mega 2560 Optiboot, anything else I can check? 

Arduino is in a second VLAN, so I use a fake Programmer... shall I switch off the NO_OTA_PORT?  -> allready tried - also no change...

**Reply by Bordman-ger** [reply]
And sorry forgot to mention, the error is not every 4th time, its by random now... the error message is "Not Found", but comes up imediate, not getting a timout after some seconds...

**Reply by JAndrassy** [reply]
2.6.4 or 2.7.4?

**Reply by Bordman-ger** [reply]
2.7.4, was only a typo sorry

**Comment by JAndrassy** [comment, marked as answer]
sorry
in src/utility/EspAtDrv.h
uncomment
//#define WIFIESPAT_MULTISERVER

I forgot to document it

**Reply by Bordman-ger** [reply]
YESSSS!!! I think that was the trick... Thanky You and dont worry (-:

**Comment by JAndrassy** [comment]
thank you for the donation. I am glad that both libraries WiFiEspAT and ArduinoOTA are useful for many

**Reply by Bordman-ger** [reply]
Yes they are.... I am afraid there could be more using it but the first steps are really hard... I started with Arduino some weeks ago, OTA with an Ethernet shield worked fine after I found a how to... but now switching to Wifi took me another 3 days of learning... for beginners like me its also not clear that a sketch on ESP is overwriting the firmware - and the trouble started with a demo sketch (-:  to understand the board and DIP settings... I will write on a little how to when my project is running. Even Jiri Bilek's Firmware is better to get started with because you do not need any seperat Firmware writer APP and can just "write a sketch"with Arduino IDE...

**Reply by Bordman-ger** [reply]
One warning I am getting, jus for your information:
In file included from D:\Projekte\Arduino\data\libraries\ArduinoOTA\src\OTAStorage.h:27:0,
                 from D:\Projekte\Arduino\data\libraries\ArduinoOTA\src\OTAStorage.cpp:1:
D:\Projekte\Arduino\data\libraries\ArduinoOTA\src\OTAStorage.cpp: In constructor 'OTAStorage::OTAStorage()':
D:\Daten\ArduinoData\packages\arduino\hardware\avr\1.8.6\cores\arduino/Arduino.h:124:26: warning: left shift count >= width of type [-Wshift-count-overflow]
 #define bit(b) (1UL << (b))
                          ^
D:\Projekte\Arduino\data\libraries\ArduinoOTA\src\OTAStorage.cpp:69:20: note: in expansion of macro 'bit'
   if (!(highBits & bit(FUSE_BOOTRST))) {
                    ^~~
